### ISSUE TITLE
Empty window from iStat Menus is showing in AltTab

### ISSUE THREAD

**Issue by SoPat712** [lwouis/alt-tab-macos#390]
**Describe the bug**

IStatMenus(the menubar app) shows as a process in alt-Tab, when it should not. Can you add a blacklist feature that just deals with this by removing it from the array, z

**Screenshots / video**

![image](https://user-images.githubusercontent.com/30350506/85960989-6b41ad80-b975-11ea-866f-4e15f758fa6f.png)

**Steps to reproduce the bug**

Download IStatMenus.

Use IStatMenus

See IStatMenus in navigation bar

**Possible Solution**

Create a blacklist feature that just removes something from the array or whatever you use to hold the apps.

**Comment by lwouis** [comment]
Hi @SoPat712! Thanks for reporting this issue!

I'm surprised to see your screenshot, as I've been using iStat Menus for years, and it has always been running on my machine during the development of AltTab. On my machine, the issue doesn't happen. I have iStats Menus 6.40, on macOS 10.15.5.

Could you please use the built-in feedback form in AltTab, and send the link to this ticket in the message box? That will open a new ticket here on github, with a debug profile of your machine (os version, AltTab version, etc). This should help investigate why the issue happens on your machine, but not mine.

**Comment by lwouis** [comment]
Also the empty window has the "hidden" icon. Have you perhaps added iStat Menus as a Login Items (i.e. `System Preferences > Users & Groups > Login Items`) with the "Hide" checkbox checked?

Also what happens when you open the main iStat Menus window? Could you share a screenshot of that?

**Comment by lwouis** [comment]
From #392:

> It's not showing up right now, but It happens like 1/2 times when I start up my macbook

<details>
<summary><strong>Debug profile</strong></summary>
<p>

* App version: 4.1.4
* App preferences: 
  * NSStatusItem Preferred Position Item-0: 20429.5
  * NSWindow Frame com.sindresorhus.Preferences.FrameAutosaveName: 485 570 340 267 0 0 1440 877 
  * SUHasLaunchedBefore: 1
  * SULastCheckTime: 2020-06-28 21:37:17 +0000
  * fadeOutAnimation: false
  * hideSpaceNumberLabels: true
  * holdShortcut: ⌘
  * mouseHoverEnabled: false
  * preferencesVersion: 4.1.4
  * showTabsAsWindows: true
* Applications: 63
* Windows: 3
  * {isFullscreen: false, isMinimized: false, isHidden: false, isTabbed: false, isOnAllSpaces: false, spaceId: 5, spaceIndex: 2}
  * {isFullscreen: false, isMinimized: false, isHidden: false, isTabbed: false, isOnAllSpaces: false, spaceId: 1, spaceIndex: 1}
  * {isFullscreen: false, isMinimized: false, isHidden: false, isTabbed: false, isOnAllSpaces: false, spaceId: 1, spaceIndex: 1}
* OS version: Version 10.15.5 (Build 19F101)
* OS architecture: x86_64
* Locale: en_US (current)
* Spaces: 2
* Dark mode: Dark
* "Displays have separate Spaces": checked
* Hardware model: MacBookAir6,2
* Screens: 2
  * {x: 0.0, y: 0.0, width: 1440.0, height: 900.0}
  * {x: -1280.0, y: 277.0, width: 1280.0, height: 1024.0}
* CPU model: Intel(R) Core(TM) i5-4250U CPU @ 1.30GHz
* Memory size: 8.59 GB
* Active CPU count: 4
* Current CPU frequency: 1.3 Ghz
* Resource utilization: 
  * CPU: 0.0%
  * Memory: 33M
  * Threads count: 17

</p>
</details>

**Comment by lwouis** [comment]
@SoPat712 oh! It's very important to learn that it's happening only when the macbook starts! I worked on a similar issue in the past in #292. I discovered some bugs in macOS on apps that launch at startup. I made multiple [workarounds](https://github.com/lwouis/alt-tab-macos/commit/2e6364418a0c2c975d6694dc41da66e730161f1a) but I guess in some scenarios it can still be an issue.

These are a pain to debug because I can't really simulate startup except by literally login out then login in. It's quite painful, and also I can't have the app with debugger at that time, so I need to work off log lines.

**Comment by SoPat712** [comment]
Sorry about the late response. This is my startup items in a screenshot

![image](https://user-images.githubusercontent.com/30350506/86165581-7c0f3200-bae1-11ea-89c3-9feb2c752269.png)

It breaks 1/2 times on startup, but then I just kill alt-tab with an alfred workflow.

Shown here:

![image](https://user-images.githubusercontent.com/30350506/86165659-9ea14b00-bae1-11ea-953d-749575d0b0a2.png)

I restart it, also with alfred:

![image](https://user-images.githubusercontent.com/30350506/86165693-a95be000-bae1-11ea-97c3-db2ebfe79f89.png)


Then iStat Menus goes away and I can continue with my use.

**Comment by lwouis** [comment]
Hey @SoPat712! I see you closed the issue. Previously you wrote:

> It breaks 1/2 times on startup

My understanding thus was that it was an actual issue, as shown on your screenshot, but that was inconsistent in its occurrence, due to startup race conditions between the different login items. Has anything changed that made you close the ticket?

If the issue can manifest itself, even inconsistently, I would like to re-open the ticket.

**Comment by SoPat712** [comment]
Hi @lwouis , since there was an easy fix to this problem(which was uncommon enough as is), I felt that there was no need to bother you with unnecessary problems that can't be reproduced. Thanks for the concern, but please, devote your time to something more important than this random error.

**Comment by lwouis** [comment]
I'll re-open the issue then 👍

**Comment by SoPat712** [comment]
I wish I could help you out in the coding process, but sadly I don't know Swift at all. One of these days I'm going to have to start learning it.

**Comment by lwouis** [comment]
Hey @SoPat712! I was curious, is the issue still happening on recent versions of AltTab these days?

**Comment by SoPat712** [comment]
Oops. I completely forgot to close this issue. The bug is not happening anymore, and I've also blacklisted iStatMenuWidget.

**Comment by lwouis** [comment]
I'll keep it open to remember to test it myself. The other day I think I saw some funky stuff at launch. I don't reboot a lot, so I don't run into these scenarios much, and I would like to take some time to actually test this more.

If you blacklisted iStats, then the issue could still be there but you are personally no longer impacted by it, whereas new users could be

**Comment by SoPat712** [comment]
I only blacklisted it in 6.2.0, I hadn't seen it after updating from 5.3 I believe. I'm pretty sure it's fixed, but if you'd like to keep it open, that's okay too.

**Comment by lwouis** [comment]
@SoPat712 Did you get this issue again since we last discussed it? I've had to logout>login many time for another issue, and I never once saw iStats show like in this ticket. Not any other app actually. I think it may be time to close this ticket.

**Comment by SoPat712** [comment]
As said before, I blacklisted the issue. It has never happened for me, but it is inconvenient to not be able to see the settings window.

**Comment by lwouis** [comment]
I added a check on window size. It will be present in the next release. I think it would prevent showing such 0 size window as we see on your screenshot above. You could un-blacklist the app, once you are on the new release.

**Comment by lwouis** [comment]
I think the command `open -j` recreates the behavior of an app launched at login, with `Login Items` > `Hide` checked. Note that you also need this to get the full behavior:

![image](https://user-images.githubusercontent.com/106195/108735001-615b6300-7573-11eb-84c4-3d35c7c13628.png)

Through that nice reproduction tool, I was able to study the inner mechanisms of this corner-case of macOS. An app launched like this will:

* Not send the expected events about creating its windows
* Not show any windows when asked for them, right after launch
* Send some `windowsFocusedChanged`/`mainWindowChanged` events quickly after launch
* Not report as `isHidden` at launch, even though it is visually hidden, and the UI says "Hide" in `Login Items` and the docs say "Hidden" in `open -j` man

Thus the workaround I added is to check if an app is `isActive` at launch. If not, we consider it has never been shown so far. Later, when if we receive `windowsFocusedChanged`/`mainWindowChanged` events, we check if the app has never been shown so far. If yes, we get its windows (at this point, we get them; unlike right after launch).

I think this workaround is going to help with Login Items, as well as maybe some other edge-cases 👍
